src/compiler/checker.ts

```diff
@@ -28417,6 +28417,96 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function checkObjectLiteral(node: ObjectLiteralExpression, checkMode?: CheckMode): Type {
+        const contextualType = getContextualType(node, /*contextFlags*/ undefined);
+        const isContextualTypeDependent =

```

Comment:
Have you tried using `couldContainTypeVariables`? A comment for this function mentions this:
```
    // Return true if the given type could possibly reference a type parameter for which
    // we perform type inference (i.e. a type parameter of a generic function). We cache
    // results for union and intersection types for performance reasons.
```

It sounds like **maybe** you could reuse it

Reply:
No because here we don't want to check if `X` has any type parameter, we want to check if `X` has a `T` type parameter. So using `isContextualTypeDependent = couldContainTypeVariable(contextualType.immediateConstraint)` would make it `true` for even `T extends F<U>` (`U` being some other type parameter) when it should have been `false` (as it's not `T extends F<T>`).

Reply:
Right, but then perhaps that function could be extended with an optional argument, and this way you could limit the results to be based on the specific type param.

Reply:
I thought about that but I think that's quite non-trivial because it's implemented keeping in mind that we're checking against any type parameter... So for example caching is a big part of this, right now because we're checking for any type parameter the cache is implemented simply as a `CouldContainTypeVariables` flag, but if we want to cache for `couldContainTypeVariables(type, typeParameters)` or even `couldContainTypeVariable(type, typeParameter)` it would require a lot of ceremony.

Reply:
Btw I realised the implementation is seriouly true to it's name "couldContainTypeVariable" and not "containsTypeVariable". This function doesn't definitively checks if the type as a type parameter, it just kinda bets, for example if it's an anonymous object type it just bets it has a type parameter and doesn't check furthur inside it. So if it returns `true` that means the type either has a type parameter or doesn't have a type parameter, but if it return `false` then it the type definately doesn't have a type parameter. So one can write bailouts like `if (!couldContainTypeVariable(type)) return` can gain some optimization and still not give bad results.

So there's no way I can use this function here as I want a definitive result. And extending it's implementation to give a definitive result might not be trivial and definitately more complex and slow than the ast check I'm doing here.

Fun fact: While working on #52088 I overlooked the "could" in it's name knowing perfectly what it means and relied on it's result which produced bad results, finally when I looked at the implementation I saw that it really takes the "could" in it's name seriously :P